share/availability/discovery/discovery.go

```diff
@@ -183,24 +209,42 @@ func (d *Discovery) ensurePeers(ctx context.Context) {
 			log.Info("Context canceled. Finishing peer discovery")
 			return
 		case <-t.C:
-			if uint(d.set.Size()) == d.peersLimit {
-				// stop ticker if we have reached the limit
-				t.Stop()
-				continue
-			}
-			peers, err := d.disc.FindPeers(ctx, topic)
-			if err != nil {
-				log.Error(err)
-				continue
-			}
-			for p := range peers {
-				go d.handlePeerFound(ctx, topic, p)
-			}
+			d.findPeers(ctx)
 		}
 	}
 }
 
+func (d *Discovery) findPeers(ctx context.Context) {
+	if uint(d.set.Size()) >= d.peersLimit {
+		log.Debugw("at peer limit, skipping FindPeers", "size", d.set.Size())
+		return
+	}
+
+	findCtx, findCancel := context.WithCancel(ctx)
+	defer findCancel()
+
+	peers, err := d.disc.FindPeers(findCtx, topic)
+	if err != nil {
+		log.Warn(err)
+		return
+	}
+
+	// we use errgroup as it obeys the context
+	wg, findCtx := errgroup.WithContext(ctx)
+	for p := range peers {
+		peer := p
+		wg.Go(func() error {
+			// pass the cancel so that we cancel FindPeers when we connected to enough peers
+			d.handlePeerFound(findCtx, peer, findCancel)
+			return nil
+		})
+	}
+	// we expect no errors
+	_ = wg.Wait()
```

Comment:
Casual `sync.Waitgroup` seems to be sufficient here, since `errgroup` never gets an error.

Reply:
I explained the reason for the errorgroup in the comment above

Reply:
I don't get it. There is no error inside error group callback. If you only need to wait for all routines to finish, why not use `sync.Waitgroup`?

Reply:
> 	// we use errgroup as it obeys the context

The regular WaitGroup does not do this